File: app/src/main/java/com/example/coroutinedemo/MainActivity.java
package com.example.coroutinedemo;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private Fragment fragName;
    private Fragment fragAge;
    private FragOutput fragResult;
    private String name;
    private String age;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        fragName = new FragName().setLabel("Your name?");
        fragAge = new FragAge().setLabel("Your age?");
        fragResult = new FragOutput();
        switchToFragment(fragName, "Get Name Fragment.");
    }

    public void switchToFragment(Fragment fragment, String fragName) {
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.root, fragment, fragName)
                .commit();
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setAge(String age) {
        this.age = age;
    }

    abstract public static class FragInput extends Fragment {
        private TextView labelTextView;

        @Nullable
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            final View view = inflater.inflate(R.layout.frag_input, container, false);
            labelTextView = view.findViewById(R.id.label);
            labelTextView.setText(getLabel());
            Button button = view.findViewById(R.id.button);
            button.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    EditText editText = view.findViewById(R.id.content);
                    onButtonClicked(editText.getText().toString());
                }

            });
            return view;
        }

        abstract public void onButtonClicked(String content);

        public static String KEY_LABEL = "The text to be shown in the label.";
        public String getLabel() {
            Bundle bundle = getArguments();
            if (null == bundle) return "(empty)";
            String s = bundle.getString(KEY_LABEL);
            if (s == null) {
                return "(empty)";
            } else {
                return s;
            }
        }

        public FragInput setLabel(String label) {
            Bundle b = getArguments();
            if (null == b) {
                b = new Bundle();
            }
            b.putString(KEY_LABEL, label);
            setArguments(b);
            if(null != labelTextView) labelTextView.setText(label);
            return this;
        }
    }

    public static class FragName extends FragInput {
        @Override
        public void onButtonClicked(String content) {
            getMainActivity().setName(content);
            getMainActivity().switchToFragment(getMainActivity().fragAge, "Get Age Fragment.");
        }

        private MainActivity getMainActivity() {
            return (MainActivity)getActivity();
        }
    }

    public static class FragAge extends FragInput {
        @Override
        public void onButtonClicked(String content) {
            getMainActivity().setAge(content);
            getMainActivity().fragResult.setLabel("Name: " + getMainActivity().name + ", Age: " + getMainActivity().age);
            getMainActivity().switchToFragment(getMainActivity().fragResult, "Show result");
        }

        private MainActivity getMainActivity() {
            return (MainActivity)getActivity();
        }
    }

    public static class FragOutput extends Fragment {
        final static public String KEY_LABEL = "The label of the output fragment";
        private TextView textView;
        @Nullable
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            View view = inflater.inflate(R.layout.frag_output, container, false);
            textView = view.findViewById(R.id.content);
            textView.setText(getLabel());
            return view;
        }

        private String getLabel() {
            Bundle b = getArguments();
            if (null == b) return "(empty)";
            String s = b.getString(KEY_LABEL);
            if (null == s) {
                return "(empty)";
            } else {
                return s;
            }
        }

        public FragOutput setLabel(String label) {
            Bundle b = getArguments();
            if (null == b) {
                b = new Bundle();
            }
            b.putString(KEY_LABEL, label);
            setArguments(b);
            if(null != textView) textView.setText(label);
            return this;
        }
    }
}

File: app/src/main/java/com/example/coroutinedemo/MainActivity.kt
package com.example.coroutinedemo

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import kotlinx.coroutines.Dispatchers
import java.lang.RuntimeException
import kotlin.coroutines.*

class MainActivity : AppCompatActivity() {

    private var fragName: Fragment? = null
    private var fragAge: Fragment? = null
    private var fragResult: FragOutput? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        fragName = FragName().setLabel("Your name?")
        fragAge = FragAge().setLabel("Your age?")
        fragResult = FragOutput()

        val complete = object: Continuation<Unit> {
            override val context: CoroutineContext
                get() = Dispatchers.Main

            override fun resumeWith(result: Result<Unit>) {
                Log.v(TAG, "The coroutine completed.")
                if (result.isFailure) {
                    result.exceptionOrNull()?.apply { throw this }
                            ?: throw RuntimeException("Suspend function failure.")
                }
            }
        }

        suspend {
            Log.v(TAG, "Already in the coroutine. Now in thread ${Thread.currentThread()}")
            coRunner()
        }.startCoroutine(complete)

        Log.v(TAG, "After starting the coroutine. Now in thread ${Thread.currentThread()}")
    }

    fun switchToFragment(fragment: Fragment?, fragName: String) {
        supportFragmentManager.beginTransaction()
                .replace(R.id.root, fragment!!, fragName)
                .commit()
    }

    abstract class FragInput : Fragment() {
        private var labelTextView: TextView? = null
        val label: String
            get() {
                val bundle = arguments ?: return "(empty)"
                val s = bundle.getString(KEY_LABEL)
                return s ?: "(empty)"
            }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val view = inflater.inflate(R.layout.frag_input, container, false)
            labelTextView = view.findViewById(R.id.label)
            labelTextView!!.text = label
            val button = view.findViewById<Button>(R.id.button)
            button.setOnClickListener {
                val editText = view.findViewById<EditText>(R.id.content)
                onButtonClicked(editText.text.toString())
            }
            return view
        }

        abstract fun onButtonClicked(content: String)

        fun setLabel(label: String): FragInput {
            var b = arguments
            if (null == b) {
                b = Bundle()
            }
            b.putString(KEY_LABEL, label)
            arguments = b
            if (null != labelTextView) labelTextView!!.text = label
            return this
        }

        companion object {

            var KEY_LABEL = "The text to be shown in the label."
        }
    }

    class FragName : FragInput() {

        private val mainActivity: MainActivity
            get() = activity!! as MainActivity

        override fun onButtonClicked(content: String) {
            mainActivity.doneGetName(content)
        }
    }

    class FragAge : FragInput() {

        private val mainActivity: MainActivity
            get() = activity!! as MainActivity

        override fun onButtonClicked(content: String) {
            mainActivity.doneGetAge(content)
        }
    }

    class FragOutput : Fragment() {
        private var textView: TextView? = null

        private val label: String
            get() {
                val b = arguments ?: return "(empty)"
                val s = b.getString(KEY_LABEL)
                return s ?: "(empty)"
            }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val view = inflater.inflate(R.layout.frag_output, container, false)
            textView = view.findViewById(R.id.content)
            textView!!.text = label
            return view
        }

        fun setLabel(label: String): FragOutput {
            var b = arguments
            if (null == b) {
                b = Bundle()
            }
            b.putString(KEY_LABEL, label)
            arguments = b
            if (null != textView) textView!!.text = label
            return this
        }

        companion object {
            val KEY_LABEL = "The label of the output fragment"
        }
    }

    var contGetName: Continuation<String>? = null
    fun doneGetName(name: String) {
        Log.v(TAG, "The name is gotten.")
        contGetName!!.apply {
            contGetName = null
            resume(name)
        }
    }

    var contGetAge: Continuation<String>? = null
    fun doneGetAge(age: String) {
        Log.v(TAG, "The age is gotten.")
        contGetAge!!.apply {
            contGetAge = null
            resume(age)
        }
    }

    private suspend fun coRunner() {
        Log.v(TAG, "Before getting name.")
        val name = suspendCoroutine<String> {
            cont -> contGetName = cont
            switchToFragment(fragName, "Get Name Fragment.")
        }
        Log.v(TAG, "Before getting age.")
        val age = suspendCoroutine<String> {
            cont -> contGetAge = cont
            switchToFragment(fragAge, "Get Age Fragment.")
        }
        Log.v(TAG, "Before showing results.")
        fragResult!!.setLabel("Name: $name, Age: $age")
        switchToFragment(fragResult, "Show result")
    }

    companion object {
        val TAG = MainActivity::class.simpleName
    }
}
